packages/ilib-po/src/Parser.ts

```diff
@@ -227,42 +276,81 @@ class Parser {
+                                                datatype: datatype ?? defaultDataType,
+                                                context: context,
+                                                index: resourceIndex++,
+                                                targetLocale: this.targetLocale?.getSpec(),
+                                                target: []
+                                            });
+                                            arrays[key] = res;
+                                        }
+                                        // add the source and target to the array at the right index
+                                        res.source[arrayIndex] = source;
+                                        if (translation) {
+                                            if (!res.target) {
+                                                res.target = [];
+                                            }
+                                            res.target[arrayIndex] = translation;
+                                        }
+                                    } else {
+                                        // no key
+                                        throw new SyntaxError(this.pathName, tokenizer.getContext());
+                                    }
                                 } else if (source) {
-                                    key = makeKey("string", source, this.contextInKey ? context : undefined);
+                                    if (!key) {
+                                        key = makeKey("string", source, this.contextInKey ? context : undefined);
+                                    }
                                     res = new ResourceString({
-                                        project: this.projectName,
+                                        project: project ?? defaultProject,
                                         key: key,
                                         sourceLocale: this.sourceLocale.getSpec(),
                                         source: source,
                                         pathName: original,
                                         state: "new",
                                         comment: comment && JSON.stringify(comment),
-                                        datatype: this.datatype,
+                                        datatype: datatype ?? defaultDataType,
                                         context: context,
                                         index: resourceIndex++,
                                         targetLocale: translation && this.targetLocale?.getSpec(),
                                         target: translation
                                     });
                                 }
-                                set.add(res);
+                                if (res) {
+                                    set.add(res);
+                                }
```

Comment:
I think it would be more correct to do
```ts
let res: Resource;
// ...
set.add(res);
```
This way, TS will evaluate all conditional branches, and it will error as long as some set of conditions can lead to `res` not being initialized (like e.g. when `!source` but `!!sourcePlurals` but accidentally `type==='array'`).

Reply:
Fixed